src/System.Buffers.Primitives/System/Buffers/ReadOnlyBufferReader.cs

```diff
@@ -94,36 +98,143 @@ private void MoveNext()
             }
         }
+
+        public void SkipTo(Position position)
+        {
+            _currentPosition = position;
+            _nextPosition = position;
+            if (_sequence.TryGet(ref _nextPosition, out ReadOnlyMemory<byte> memory))
+            {
+                _currentSpan = memory.Span;
+            }
+            else
+            {
+                _currentSpan = default;
+            }
+            _index = 0;
+        }
+      
+        public Position? PositionOf(byte value)
+        {
+            var unread = UnreadSegment;
+            var index = unread.IndexOf(value);
+            if (index != -1) return _currentPosition + (_index + index);
+
+            var nextSegmentPosition = _nextPosition;
+            var currentSegmentPosition = _nextPosition;
+            while (_sequence.TryGet(ref nextSegmentPosition, out var memory, true))
+            {
+                var segment = memory.Span;
+                index = segment.IndexOf(value);
+                if (index != -1) return currentSegmentPosition + index;
+                currentSegmentPosition = nextSegmentPosition;
+            }
+            return default;
+        }
+
+        public Position? PositionOf(ReadOnlySpan<byte> value)
+        {
+            var unread = UnreadSegment;
+            var index = unread.IndexOf(value);
+            if (index != -1) return _currentPosition + (_index + index);
+            if (value.Length == 0) return default;
+
+            Span<byte> temp = stackalloc byte[(value.Length - 1) * 2];
+            var currentSegmentPosition = _currentPosition;
+            var nextSegmentPosition = _nextPosition;
+            var currentSegmentConsumedBytes = _index;
+
+            while (true)
+            {
+                // Try Stitched Match
+                int tempStartIndex = unread.Length - Math.Min(unread.Length, value.Length - 1);
+                var candidatePosition = currentSegmentPosition + (currentSegmentConsumedBytes + tempStartIndex);
+                int copied = CopyFromPosition(_sequence, candidatePosition, temp);
+                if (copied < value.Length) return null;
+
+                index = temp.Slice(0, copied).IndexOf(value);
+                if (index < value.Length)
+                {
+                    if (index != -1) return candidatePosition + index;
+                }
+
+                currentSegmentPosition = nextSegmentPosition;
+                // Try Next Segment
+                if (!_sequence.TryGet(ref nextSegmentPosition, out var memory, true))
+                {
+                    return default;
+                }
+                currentSegmentConsumedBytes = 0;
+                unread = memory.Span;
+
+                index = unread.IndexOf(value);
+                if (index != -1) return currentSegmentPosition + index;
+            }
+        }
+
+        public static int CopyTo(BufferReader<TSequence> bytes, Span<byte> buffer)
```

Comment:
There is same logic on `ROB`. Lets not duplicate it.

Reply:
Yeah, but I don't have access to ROB. I agree that it would be good to somehow refactor to reuse the logic. Any ideas for how to refactor?

Reply:
We decided some time ago that we won't try to implement everything on top of `ISequence` and now we are trying to do it.

We should expose `Sequence` and so allow access to all these implementations on `ROB`. We can expand ISequence in the future to accommodate new features but we shouldn't go into rabbit hole right now. Especially considering these implementations are not used in production code and may not be thoroughly tested.

Reply:
We can expose `SliceBefore()` and `SliceAfter()` extension methods in the way I talked in other thread so the consumer can get `ROB` of either and call existing methods on it.

Reply:
I am not sure how it helps here. I don't have BufferReader<ROB>; I have BufferReader<TSequence>

Reply:
As I said, in a previous meeting we decided not to try building everything on top of `ISequence`, especially while we have only one real implementation. We can live with supporting these things only on ROB and see what other things would implement `ISequence<ReadOnlyMemory<byte>>` if there would be any other at all.

Reply:
Maybe moving the reader to `ISequence` was a mistake too.

Reply:
We have several problems in various abstractions: ISequence, Position, ROB. But I would not throw baby with the bathwater. I implemented many extensions over the APIs we have now (the no merge PR) and it seems to be almost working. I think we need to solve a couple of small problems to enable both reading over sequences and slicing.